### PR TITLE
Service push notifications

### PR DESCRIPTION
**Pull request by EvsenevDev**
- **AppDelegate** rewritten to Swift
- Added **Push Notification Service**
- Added **saveDevice(token:)** to AuthService for upload FCM token

Push Notification Service:
- Handle Notifications and unwrap JSON then emit it into **notificationJSON** observable.
- Handle and Send FCM token to **users/me**

### CODE REVIEW

File: SmartReceipts/AppDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        
+        Logger.info("Language: \(Locale.preferredLanguages.first!)")
+        
+        NSSetUncaughtExceptionHandler { exception in
+            RateApplication.sharedInstance().markAppCrash()
+            
+            var message = exception.description
+            message += "\n"
+            message += exception.callStackSymbols.description
+            Logger.error(message, file: "UncaughtExcepetion", function: "onUncaughtExcepetion", line: 0)
```
Comment (wbaumann) [inline SmartReceipts/AppDelegate.swift]: Will this still work for Obj-C crash stacks? I assume the NSString -> String conversions will still work

File: SmartReceipts/Services/PushNotificationService.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+//  Copyright © 2017 Will Baumann. All rights reserved.
+//
+
+import Foundation
+import UserNotifications
+import RxSwift
+import FirebaseMessaging
+import SwiftyJSON
+
+class PushNotificationService: NSObject {
```
Comment (wbaumann) [inline SmartReceipts/Services/PushNotificationService.swift]: Is it easy to add tests for this class too?

File: SmartReceipts/AppDelegate.swift (hunk 1 above)
Comment (EvsenevDev) [reply SmartReceipts/AppDelegate.swift]: @wbaumann yes

File: SmartReceipts/Services/PushNotificationService.swift (hunk 2 above)
Comment (EvsenevDev) [reply SmartReceipts/Services/PushNotificationService.swift]: @wbaumann for test it i need handle Push from FCM. Do you have idea how test it? I have one:
Call Delegate method with my UserData if it possible and try to subscribe to observable. What do you think?

Comment (wbaumann) [reply SmartReceipts/Services/PushNotificationService.swift]: @EvsenevDev Yea - I think that'd be good. Fake being logged in and verify that it calls your mock service on subscribe

**Pull request merged**